vm ctrl: Allow live changes to network fields

orelmisan opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
Currently, the "RestartRequired" condition is added to the VirtualMachine object for every change made to network-related fields.
This is true for both `LiveUpdate` and `Staging` rollout strategies.

Users of the NIC hotplug / hotunplug and link state management features are required to restart their VMs in order for the "RestartRequired" condition to be removed.

Allow the following changes to network-related fields:
- Addition of interfaces and networks (hotplug).
- Changing the interface's state (hotunplug and link state management).

All other changes that are not explicitly mentioned - require restarting the VM in order for them to be applied.

The newly added logic does not depend on the rollout strategy, due to the following reasons:
 - NIC hotplug and hotunplug can be performed in-place.
 - Link state management is performed in-place.

Note: There is a webhook in-place that verifies that interfaces and networks match.

<!-- (optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: -->
Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: <!-- optional: slack, other GH issue, mailinglist, ... -->
https://github.com/kubevirt/enhancements/pull/8

### Special notes for your reviewer

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
The "RestartRequired" condition is not set on VM objects for live-updatable network fields
```

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/e2481365bd7b1e5c560dd368c49cc98fd11a5739..f9a7df61e36b9a0506ca96b7b3de67286fbae091): Changed function suffixes to `Restart` instead of `Reboot`.

orelmisan commented:
/cc @oshoval @nirdothan

lyarwood commented on pkg/network/vmispec/restart_test.go:
```diff
@@ -0,0 +1,197 @@
+			libvmi.WithInterface(libvmi.InterfaceDeviceWithMasqueradeBinding()),
+			libvmi.WithNetwork(v1.DefaultPodNetwork()),
+		)
+		vm := libvmi.NewVirtualMachine(vmi).DeepCopy()
```
Something for a follow up but I wonder if we should always `DeepCopy()` before returning the vm from `NewVirtualMachine`?

orelmisan replied:
It makes sense to me.

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/f9a7df61e36b9a0506ca96b7b3de67286fbae091..a8ffd254cbf1b3121828645f5962d88b8e3872ea): Rebase.

orelmisan commented on pkg/network/vmispec/restart_test.go:
```diff
@@ -0,0 +1,197 @@
+		secondaryNADName2 = "bar-nad"
+	)
+
+	It("should not require restart the VM when there are no networks", func() {
```
Done.

orelmisan commented on pkg/network/vmispec/restart_test.go:
```diff
@@ -0,0 +1,197 @@
+		Expect(vmispec.ShouldChangeRequireRestart(vm, vmi)).To(BeFalse())
+	})
+
+	It("should not require restart the VM when there are no changes", func() {
```
It was shortened to `should not require restart ` all over.

orelmisan commented on pkg/network/vmispec/restart_test.go:
```diff
@@ -0,0 +1,197 @@
+		Entry("From empty to absent", v1.InterfaceState(""), v1.InterfaceStateAbsent),
+		Entry("From empty to up", v1.InterfaceState(""), v1.InterfaceStateLinkUp),
+		Entry("From empty to down", v1.InterfaceState(""), v1.InterfaceStateLinkDown),
+		Entry("From absent to empty", v1.InterfaceStateAbsent, v1.InterfaceState("")),
```
Done.

orelmisan commented on pkg/virt-controller/watch/vm/vm.go:
```diff
@@ -2936,6 +2936,11 @@ func (c *Controller) addRestartRequiredIfNeeded(lastSeenVMSpec *virtv1.VirtualMa
 		}
 	}
 
+	if !vmispec.ShouldChangeRequireRestart(currentVM, vmi) {
```
Done.

orelmisan commented on pkg/network/vmispec/restart.go:
```diff
@@ -0,0 +1,79 @@
+/*
```
Done.

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/a8ffd254cbf1b3121828645f5962d88b8e3872ea..fe624965329dcc5a8d6b65fdb3117d8f9215b0c6): Addressed review comments.

orelmisan commented on pkg/network/vmispec/restart.go:
```diff
@@ -0,0 +1,79 @@
+	for name, lastSeenIface := range lastSeenIfacesByName {
+		currentIface, exists := currentIfacesByName[name]
+		if !exists {
+			return true
```
Hotunplug is performed by setting the interface's state fields to `absent` [1]. The network VM controller is responsible for actually removing the interfaces and networks from the spec.

[1] https://kubevirt.io/user-guide/network/hotplug_interfaces/#remove-interface

orelmisan replied:
This logic was extracted to a function (same for networks).

orelmisan commented on pkg/network/vmispec/restart.go:
```diff
@@ -0,0 +1,79 @@
+		lastSeenIfaceCopy := lastSeenIface.DeepCopy()
+		lastSeenIfaceCopy.State = ""
+
+		currentIfaceCopy := currentIface.DeepCopy()
```
I'm unsure that this approach will work, since v1.Network and v1.Interface have pointer fields.

orelmisan commented on tests/network/link_state.go:
```diff
@@ -122,6 +122,11 @@ var _ = Describe(SIG("interface state up/down", func() {
 
 		Expect(patchToggleVMInterfacesStates(vm)).To(Succeed())
 
+		Consistently(matcher.ThisVM(vm)).
+			WithTimeout(5 * time.Second).
+			WithPolling(1 * time.Second).
+			Should(matcher.HaveConditionMissingOrFalse(v1.VirtualMachineRestartRequired))
+
```
@dasionov could you please elaborate what are your concerns with regards to flakiness of the current implementation?

EdDev replied:
@dasionov I understand that you want to assure that the status is refreshed by the controller before attempting to assert, assuring something did changed in this period.
However, I think this will not assure anything. The controller may have updated the status due to a different reason between the time the test observed the prev status and the controller processing the new spec.
It is a classic race condition.

EdDev commented on pkg/virt-controller/watch/vm/vm.go:
```diff
@@ -2936,6 +2936,11 @@ func (c *Controller) addRestartRequiredIfNeeded(lastSeenVMSpec *virtv1.VirtualMa
+		lastSeenVM.Spec.Template.Spec.Networks = currentVM.Spec.Template.Spec.Networks
+	}
+
 	if !equality.Semantic.DeepEqual(lastSeenVM.Spec.Template.Spec, currentVM.Spec.Template.Spec) {
```
Based on this, if the order is changed + a non-reset change ( e.g. `state` or hotplug) on the new VM, `!netvmispec.ShouldChangeRequireRestart` will resolve to `false` but here we will resolve to `true` and therefore mark as restart-required which seems wrong to me.

We need the e2e test to cover this, because I do not see how this can be covered in unit tests at the moment.

orelmisan replied:
Good catch!
IIUC, SIG storage have the same problem @alicefr.

A test was added on our side to check that we don't consider the interfaces and networks ordering.

> We need the e2e test to cover this, because I do not see how this can be covered in unit tests at the moment.

This is still an open point.

EdDev replied:
Looking at this again, there is no actual problem except a coverage one which has been taken care of by unit testing the required-restart logic. If a reset is not required, the network spec is synced so it will not effect the equality check here.

The e2e are needed only if we cannot unit-test the overall logic here.

EdDev reviewed (changes requested):
Reviewed without the last commit (e2e test).

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/fe624965329dcc5a8d6b65fdb3117d8f9215b0c6..6d40610cfc0dc2ceb1a2c14e0ec1b933213b5e0d): Partially addressed @EdDev's [comments](https://github.com/kubevirt/kubevirt/pull/14602#pullrequestreview-2817918681).

orelmisan commented on pkg/network/vmispec/restart.go:
```diff
@@ -0,0 +1,79 @@
+
+func ShouldChangeRequireRestart(vm *v1.VirtualMachine, vmi *v1.VirtualMachineInstance) bool {
+	lastSeenIfaces := vmi.Spec.Domain.Devices.Interfaces
+	currentIfaces := vm.Spec.Template.Spec.Domain.Devices.Interfaces
```
The VM interfaces and networks are now referred to as `desired` and the VMI's as `current`.

orelmisan commented on pkg/network/vmispec/restart.go:
The logic was moved under pkg/network/restart and the function was renamed to `IsRequired`.

orelmisan commented on pkg/network/vmispec/restart.go:
```diff
@@ -0,0 +1,79 @@
+			return true
+		}
+
+		lastSeenIfaceCopy := lastSeenIface.DeepCopy()
```
The variables were renamed to `normalizedIface1` and `normalizedIface2`.

orelmisan commented on pkg/network/restart/restart_suite_test.go:
```diff
@@ -0,0 +1,13 @@
+package restart_test
```
Thanks, done.

orelmisan commented on pkg/network/restart/restart_test.go:
```diff
@@ -0,0 +1,213 @@
+		)
+		vm := libvmi.NewVirtualMachine(vmi)
+
+		Expect(restart.IsRequired(vm, vmi)).To(BeFalse())
```
The first two tests were combined into a table.
I think the third (hotplug scenario) deserves its own test.

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/09ba68f74f1cfe5bb8322a1ddaf74e8645cf1681..da608ace21b17bed0f729c6ebbf1d4f6362f65aa): Address @EdDev's [comments](https://github.com/kubevirt/kubevirt/pull/14602#pullrequestreview-2819540018).

EdDev commented on tests/network/link_state.go:
```diff
@@ -122,6 +122,11 @@ var _ = Describe(SIG("interface state up/down", func() {
 		Expect(patchToggleVMInterfacesStates(vm)).To(Succeed())
 
+		Consistently(matcher.ThisVM(vm)).
+			WithTimeout(5 * time.Second).
```
I think you are a bit optimistic here. I would keep it around 15sec or even more.

There is another option, but I am unsure how the current live-update logic works.
If the "restart" is revertible, i.e. do a change, then undo it and see the "restart" condition go away, we could patch the VM so it will appear and then undo it with the link-state change and check it got removed.
Maybe too crazy, not sure. (but if it works, it is faster in terms of execution)

orelmisan replied:
> I think you are a bit optimistic here. I would keep it around 15sec or even more.

The timeout was changes to 15 seconds.

> There is another option, but I am unsure how the current live-update logic works.
If the "restart" is revertible, i.e. do a change, then undo it and see the "restart" condition go away, we could patch the VM so it will appear and then undo it with the link-state change and check it got removed.
Maybe too crazy, not sure. (but if it works, it is faster in terms of execution)

The RestartRequired condition is not revertible, i.e. it will only be removed after the VM is stopped [1] or restarted [2].

[1] https://github.com/kubevirt/kubevirt/blob/1227fe24f4a627fd9fc0472df4cfd2dcdf6735e3/pkg/virt-controller/watch/vm/vm.go#L1588
[2] https://github.com/kubevirt/kubevirt/blob/1227fe24f4a627fd9fc0472df4cfd2dcdf6735e3/pkg/virt-controller/watch/vm/vm.go#L1186

EdDev replied:
> The RestartRequired condition is not revertible, i.e. it will only be removed after the VM is stopped [1] or restarted [2].

Ohh, this is very limiting.
We should raise it with sig-compute at some forum.
Thank you for checking this.

EdDev reviewed (approved):
Thank you!

I had one concern about the short time you wait to assure the "restart" condition does not come, but I leave it to you do decide.

/approve

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/da608ace21b17bed0f729c6ebbf1d4f6362f65aa..0dc697cdf3667f0cb8ffb1b6694cf09317d72cbd): Increased Consistently timeout to 15 sec per @EdDev's [suggestion](https://github.com/kubevirt/kubevirt/pull/14602#discussion_r2076881840).

enp0s3 commented:
/approve
@orelmisan Thank you!

orelmisan commented:
/retest-required
